Infrastructure: refactor and simplify profile closing

SlySven opened the pull request:
#### Brief overview of PR changes/additions
An attempt to improve the process of closing multiple profiles, including handling the case where a closure of the whole Mudlet application is cancelled part way through by the end-user selecting the "Cancel" option when asked if they want to save a profile or not.

#### Motivation for adding to Mudlet
When multi-playing carrying out the above action (cancelling part way through) can leave widgets and dialogues for profiles open even though the underlying `Host` instance has gone away. This can easily lead to fatal segmentation faults when those widgets or dialogues try to reference something that is no longer present. Additionally the code prior to this PR had multiple execution paths when "closing" a profile leading to some of them not doing all the things they needed to - hence the issues just listed. This PR tries to rationalise the execution path and reduce the amount of duplication so that it is easier to follow and amend in the future.

#### Other info (issues closed, discussion etc)
A `closeEvent(QEvent*)` for the `TMainConsole` class has been created to override the `TConsole` one as it is the only console that has to be concerned with saving the profile.

The way the code was structured meant that the destruction of `TDockWidget`s would sometimes hit a
`"TConsole::closeEvent(QCloseEvent*) INFO - closing a UserWindow but the TDockWidget pointer was not found to be removed...";` message - this was because the relevant `TDockWidget` had already been removed from the `(QMap<QString, TDockWidget*>) TMainConsole::mDockWidgetMap` before the `TConsole::closeEvent(...)` was executed.

The `(bool) Host::mIsGoingDown` flag was unnecessary and seemed to duplicate the `(bool) Host::mIsClosingDown` one - OTOH I did need a new `(bool) Host::mForcedClose` flag to signal that the main application was being forcibly closed and that everything should be saved whether the user had set the "auto-save on exit option" or not. The `(bool) TConsole::mUserAgreedToCloseConsole` flag has been renamed and moved to be `(bool) TMainConsole::mEnableClose` instead.

The actions to close down widgets and dialogues associated with a profile have been moved to the new `(void) Host::closeChildren` method which is called by ~~both the `(void) Host::forceClose()` method - which cannot be aborted and~~ the `(bool) Host::requestClose()` which can be aborted, and returns `false` if that *has* happened. ***Edit: the forceClose() method now just sets a flag which forces the requestClose() to not ask questions and just save things...*** 

The `HostManager::deleteHost(const QString&)` return value was never used so I changed it to be a `void` method. Note that this is the method that is actually responsible for destroying the `Host` instance for each profile.

The `(static void) TDebug::removeHost(Host*)` method needs the name of the profile that is "going away" but because it is called from the `Host` destructor I ran into a copy of seg. faults from:
* calling `Host::getName()` from there
* calling `TBuffer::log(...)` indirectly from there because that dereferences `Host::mpConsole` to check if the `TBuffer` is the one for the main console and there wasn't a valid `Host` at that point.

The process of iterating through all the open profiles to close them down in `(void) mudlet::closeEvent(QCloseEvent*)` is more tricky than it seems because the `for (auto pHost : mHostManager)` loop blows up if `mHostManager` is modified in anyway during the iteration. Instead I found I had to note down the profiles that have been closed and only destroy them after the "Do you wanna save this profile?" loop has been gone through for as many as get closed.

Because the package exporter widget is parented to the widget that is effectively the `QMainWindow` for the mudlet instance rather than the `TMainConsole` for it's profile, it was not getting closed/destroyed should it have been open when the profile it was associated with get closed but the Mudlet application stay running. As the `Host` class is NOT derived from a `QWidget` the dialogue could not be made a child of it; instead I've arranged for the `QObject::destroyed` signal of the `Host` to cancel and close it instead.

SlySven committed:
Infrastructure: refactor and simplify profile closing

#### Brief overview of PR changes/additions
An attempt to improve the process of closing multiple profiles, including
handling the case where a closure of the whole Mudlet application is
cancelled part way through by the end-user selecting the "Cancel" option
when asked if they want to save a profile or not.

#### Motivation for adding to Mudlet
When multi-playing carrying out the above action (cancelling part way
through) can leave widgets and dialogues for profiles open even though the
underlying `Host` instance has gone away. This can easily lead to fatal
segmentation faults when those widgets or dialogues try to reference
something that is no longer present. Additionally the code prior to this
PR had multiple execution paths when "closing" a profile leading to some
of them not doing all the things they needed to - hence the issues just
listed. This PR tries to rationalise the execution path and reduce the
amount of duplication so that it is easier to follow and amend in the
future.

#### Other info (issues closed, discussion etc)
A `closeEvent(QEvent*)` for the `TMainConsole` class has been created to
override the `TConsole` one as it is the only console that has to be
concerned with saving the profile.

The way the code was structured meant that the destruction of `TDockWidget`s
would sometimes hit a
`"TConsole::closeEvent(QCloseEvent*) INFO - closing a UserWindow but the TDockWidget pointer was not found to be removed...";`
message - this was because the relevant `TDockWidget` had already been
removed from the `(QMap<QString, TDockWidget*>) TMainConsole::mDockWidgetMap`
before the `TConsole::closeEvent(...)` was executed.

The `(bool) Host::mIsGoingDown` flag was unnecessary and seemed to
duplicate the `(bool) Host::mIsClosingDown` one - OTOH I did need a new
`(bool) Host::mForcedClose` flag to signal that the main application was
being forcibly closed and that everything should be saved whether the user
had set the "auto-save on exit option" or not. The
`(bool) TConsole::mUserAgreedToCloseConsole` flag has been renamed and
moved to be `(bool) TMainConsole::mEnableClose` instead.

The actions to close down widgets and dialogues associated with a profile
have been moved to the new `(void) Host::closeChildren` method which is
called by both the `(void) Host::forceClose()` method - which cannot be
aborted and the `(bool) Host::requestClose()` which can, and returns
`false` if that *has* happened.

The `HostManager::deleteHost(const QString&)` return value was never
used so I changed it to be a `void` method. Note that this is the method
that is actually responsible for destroying the `Host` instance for each
profile.

The `(static void) TDebug::removeHost(Host*)` method needs the name of the
profile that is "going away" but because it is called from the `Host`
destructor I ran into a copy of seg. faults from:
* calling `Host::getName()` from there
* calling `TBuffer::log(...)` indirectly from there because that
dereferences `Host::mpConsole` to check if the `TBuffer` is the one for
the main console and there wasn't a valid `Host` at that point.

The process of iterating through all the open profiles to close them down
in `(void) mudlet::closeEvent(QCloseEvent*)` is more tricky than it seems
because the `for (auto pHost : mHostManager)` loop blows up if
`mHostManager` is modified in anyway during the iteration. Instead I found
I had to note down the profiles that have been closed and only destroy them
after the "Do you wanna save this profile?" loop has been gone through
for as many as get closed.

Because the package exporter widget is parented to the widget that is
effectively the `QMainWindow` for the mudlet instance rather than the
`TMainConsole` for it's profile, it was not getting closed/destroyed
should it have been open when the profile it was associated with get
closed but the Mudlet application stay running. As the `Host` class is NOT
derived from a `QWidget` the dialogue could not be made a child of it;
instead I've arranged for the `QObject::destroyed` signal of the `Host`
to cancel and close it instead.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: allow Lua closeMudlet() to work

Destroying the Lua interpreter in a profile whilst it is still executing
(the `closeMudlet()` function) has fatal side-effects!

Also, trying to close the Central Debug console when there isn't one is
also a bad idea (but not one I initially picked up on as I had one open
during development testing).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: remove code that calls TMainConsole::close() redundantly

Also prune an unused setter for `(bool) Host::mIsClosingDown`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/Host.cpp:
```diff
@@ -477,11 +475,112 @@ Host::~Host()
+    }
+
+    mForcedClose = true;
+    postMessage(tr("[ ALERT ] - this profile is being forced to save and close."));
```
I don't think we need additional user messages about this, Mudlet should be as least chatty as possible for the players.

@Mudlet/ui-review

SlySven replied:
Well if one profile calls `closeMudlet()` than ALL the profiles get closed with no arguments - and if there is any sort of logging to file going on the user might want to record that is why the session terminated abruptly...

Kebap replied:
I hear the explanation and now think this is valid then. 
However at least please upper the T after the dash starting a new sentence.

The "being forced" seems hard to grasp and hard to translate. How about reducing the cognitive load...
- ~This profile is being forced to save and close.~
- This profile is forced to save and close.
- This profile is asked to save and close.
- This profile was asked to save and close.
- This profile will now save and close.

Given we don't specify exactly "who" was the one being the force (and we shouldn't add that either), we can keep it shorter and simpler.
Either the player closed the profile herself (and we probably don't need a message then but hey) or it was asked by others and will now comply. Then it is okay to print that bit for future reference. If all goes well, it won't be displayed for long either. Only in log files it may accumulate for eternity, but it's small enough I guess.

```suggestion
    postMessage(tr("[ ALERT ] - This profile will now save and close."));
```

vadi2 replied:
That sounds great to me 👍

vadi2 commented on src/Host.cpp:
```diff
@@ -477,11 +475,112 @@ Host::~Host()
+    // the close() method called here to return a true if the event was
+    // accepted:
+    if (!mpConsole->close()) {
+        // Nope the user doesn't want this to close - and it won't have set it's
```
```suggestion
        // Nope the user doesn't want this to close - and it won't have set its
```

vadi2 commented on src/Host.cpp:
```diff
@@ -477,11 +475,112 @@ Host::~Host()
+        return false;
+    }
+
+    // The above will have initiated a save of the profile (and it's map) if it
```
```suggestion
    // The above will have initiated a save of the profile (and its map) if it
```

vadi2 commented on src/Host.cpp:
```diff
@@ -477,11 +475,112 @@ Host::~Host()
+void Host::closeChildren()
+{
+    mIsClosingDown = true;
+    std::list<QPointer<TToolBar>> const hostToolBarMap = getActionUnit()->getToolBarList();
```
```suggestion
    const auto hostToolBarMap = getActionUnit()->getToolBarList();
```

SlySven replied:
Fair enough - that was stuff I just cut-and-pasted...

vadi2 commented on src/Host.h:
```diff
@@ -910,6 +912,10 @@ private slots:
     // Whether to display each item's ID number in the editor:
     bool mShowIDsInEditor = false;
+
+    // Set when the mudlet singleton demands that we close - used to force a
```
```suggestion
    // Set when the mudlet singleton demands that we close - used to force an
```

SlySven replied:
Already pointed out to me and fixed.

vadi2 commented on src/HostManager.cpp:
```diff
@@ -25,16 +25,18 @@
+        return;
     }
+
+    // As this pulls the QSharedPointer that hostname identifies out of the pool
+    // the Host goes out of scope when execution leaves this method and thus
+    // gets destroyed:
```
```suggestion
    // this finally deletes the host (thanks to QSharedPointer)
```

vadi2 commented on src/TConsole.cpp:
```diff
@@ -722,14 +722,14 @@ void TConsole::closeEvent(QCloseEvent* event)
-            mudlet::self()->refreshTabBar();
-            event->ignore();
-            return;
         }
```
Nice quality improvements 👍

vadi2 commented on src/TMainConsole.cpp:
```diff
@@ -1532,3 +1532,113 @@ void TMainConsole::showStatistics()
     mpHost->mpConsole->raise();
 }
+
+void TMainConsole::closeEvent(QCloseEvent* event)
```
Great call to move this to `TMainConsole`.

vadi2 commented on src/TMainConsole.cpp:
```diff
@@ -1532,3 +1532,113 @@ void TMainConsole::showStatistics()
+    if (mpHost->mFORCE_SAVE_ON_EXIT || mpHost->isClosingForced()) {
+        mudlet::self()->saveWindowLayout();
+        mpHost->modulesToWrite.clear();
+        // We are not checking the status result from here!
```
Should we check the status and the code needs to be changed? Or this is intended that the status it not to be checked and the core is valid?

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -133,6 +133,13 @@ dlgPackageExporter::dlgPackageExporter(QWidget *parent, Host* pHost)
+    // of the profile it would otherwise be left around if only the profile was
+    // being closed:
+    connect(mpHost, &QObject::destroyed, this, &dlgPackageExporter::slot_cancelExport);
+    connect(mpHost, &QObject::destroyed, this, &dlgPackageExporter::close);
```
Nice fix!

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1803,34 +1747,53 @@ Host* mudlet::getActiveHost()
+    bool abortClose = false;
+    // Due to the way that Hosts are stored we cannot do a closeHost(hostName)
+    // within the following loop as it fatally messes with what mHostManager
+    // contains - this is STL iterator stuff! Slysven 2024/04
```
I'm not sure of the value of leaving named comments, what does naming and stamping them add? I'd much rather have everyone share the code.

SlySven replied:
This code bit me - and since it is our only (so far) use of `QScopedPointer<T>` I wanted to warn anyone tempted to frob things here - and since I found it it didn't seem unreasonable to identify myself in case anyone else wanted to ask why this comment was here without having to go git history reading...

vadi2 replied:
I don't foresee anyone taking you up on this offer, as the Qt documentation is rather straightforward here. I don't think we should start this trend...

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1803,34 +1747,53 @@ Host* mudlet::getActiveHost()
+        if (pHost->requestClose()) {
+            // If we get here then the user has agreed to close it and the
+            // profile has been saved - if required - or both have happened
+            // automatically - and the main console has been told to close:
```
👍 great comment explaining the 'why'

vadi2 commented on src/mudlet.cpp:
```diff
@@ -1803,34 +1747,53 @@ Host* mudlet::getActiveHost()
+        // This profile is not to be closed or the user has cancelled the close,
+        // in either case the application close cannot proceed - so give up,
+        // but we cannot just ignore() the event and return as there may be
+        // previously closed profiles to clean up:
```
👍 great comment explaining the 'why'

vadi2 commented on src/mudlet.cpp:
```diff
@@ -4947,3 +4860,14 @@ void mudlet::onlyShowProfiles(const QStringList& predefinedProfiles)
+// The Lua interpreter cannot call mudlet::forceClose() directly as the latter
+// will destroy the former before a direct call has completed which has bad
+// effects (like the Lua API resetProfile() once did). Instead arrange for it
+// to be done by a "zero" timer:
```
```suggestion
// to be done on the next Qt event loop iteration:
```

Kebap commented on src/Host.cpp:
```diff
@@ -842,8 +941,9 @@ std::tuple<bool, QString, QString> Host::saveProfile(const QString& saveFolder,
         // This is likely to be the save as the profile is closed
-        qDebug().noquote().nospace() << "Host::saveProfile(...) INFO - called with no saveFolder or saveName arguments "
-                                        "so assuming it is a end of session save and the TCommandLines' histories need saving...";
+        qDebug().noquote().nospace() << "Host::saveProfile(...) INFO - called with no saveFolder or saveName arguments for "
+                                     << mHostName
+                                     << " so assuming it is a end of session save and the TCommandLines' histories need saving...";
```
```suggestion
        qDebug().noquote().nospace() << "Host::saveProfile(...) INFO - called with no saveFolder or saveName arguments for profile '"
                                     << mHostName
                                     << "' so assuming it is a end of session save and the TCommandLines' histories need saving...";
```

Kebap commented on src/Host.h:
```diff
@@ -910,6 +912,10 @@ private slots:
     // Whether to display each item's ID number in the editor:
     bool mShowIDsInEditor = false;
+
+    // Set when the mudlet singleton demands that we close - used to force a
+    // attempt to save the profile and map - without asking:
```
```suggestion
    // Set when the mudlet singleton demands that we close - used to force an
    // attempt to save the profile and map - without asking:
```

Kebap commented:
Hooray for you to rationalise the execution path and reduce the amount of duplication. Great job!

SlySven committed:
Revise: apply suggestions from code review

Co-authored-by: Kebap <kebap_spam@gmx.net>

vadi2 commented on src/Host.cpp:
```diff
@@ -477,11 +475,112 @@ Host::~Host()
     mErrorLogStream.flush();
     mErrorLogFile.close();
-    TDebug::removeHost(this);
+    // We are in the destructor - so it is not a good idea for anything we call
+    // to refer to our members, given that we want the name of the profile in
+    // the following pass it in as a value so we don't need to access the member
```
I find this hard to parse, how about this instead?

```suggestion
    // Since this is a destructor, it's risky to rely on member variables within the destructor itself.
    // To avoid this, we can pass the profile name as an argument instead of accessing it
    // directly as a member variable. This ensures the destructor doesn't depend on the 
    // object's state being valid.

```

vadi2 reviewed (commented):
Great work untangling this organic growth 👍

mpconley commented on src/Host.cpp:
```diff
@@ -842,8 +941,9 @@ std::tuple<bool, QString, QString> Host::saveProfile(const QString& saveFolder,
-                                        "so assuming it is a end of session save and the TCommandLines' histories need saving...";
+        qDebug().noquote().nospace() << "Host::saveProfile(...) INFO - called with no saveFolder or saveName arguments for profile '"
+                                     << mHostName
+                                     << "' so assuming it is a end of session save and the TCommandLines' histories need saving...";
```
``
```suggestion
                                     << "' so assuming it is an end of session save and the TCommandLines' histories need saving...";
```

mpconley commented:
Looks good. Tested with single & multiple profiles.

SlySven committed:
Apply suggestions from peers' code review

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>
Co-authored-by: Mike Conley <sousesider@gmail.com>

SlySven committed:
Update src/mudlet.cpp

Somehow that got missed when I was amalgamating all the other peer review suggestions.

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

SlySven commented on src/mudlet.cpp:
```diff
@@ -1803,34 +1747,53 @@ Host* mudlet::getActiveHost()
+    bool abortClose = false;
+    // Due to the way that Hosts are stored we cannot do a closeHost(hostName)
+    // within the following loop as it fatally messes with what mHostManager
+    // contains - this is STL iterator stuff! Slysven 2024/04
```
```suggestion
    // contains - this is STL iterator stuff!
```

SlySven committed:
Update src/mudlet.cpp

Remove an identifier for me with a date.

SlySven committed:
Merge branch 'development' into Infrastructure_refactorAndSimplifyProfileClosing

SlySven commented:
/refresh links

SlySven committed:
Update src/Host.cpp

Peer review suggestion applied.

Co-authored-by: Kebap <kebap_spam@gmx.net>

SlySven committed:
Merge branch 'development' into Infrastructure_refactorAndSimplifyProfileClosing
